### PR TITLE
feat(charts): add log forwarding via Fluent Bit sidecar

### PR DESCRIPTION
**Pull request by SanyaKochhar**
**Description**: Adds functionality for optionally enabling a Fluent Bit sidecar to allow users to forward logs to an output of choice

**Affected area**:

- New Functionality      [x]
- Documentation          [x]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

    - [Inspiration](https://samcogan.com/export-kubernetes-logs-to-azure-log-analytics-with-fluent-bit/) for Fluent Bit to Log Analytics example

### CODE REVIEW

File: charts/osm/templates/osm-deployment.yaml [hunk 1]
```diff
@@ -73,6 +73,33 @@ spec:
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.name
+      {{- if .Values.OpenServiceMesh.enableFluentbit}}
```
Comment (shashankram) [inline charts/osm/templates/osm-deployment.yaml:76]:
```suggestion
      {{- if .Values.OpenServiceMesh.enableFluentbit }}
```

File: charts/osm/templates/osm-deployment.yaml [hunk 2]
```diff
@@ -73,6 +73,33 @@ spec:
...
+            mountPath: /fluent-bit/etc
+            readOnly: true 
+          - name: var-log-containers
+            mountPath: /var/log/containers
+            readOnly: true
+          - name: var-log-pods
+            mountPath: /var/log/pods
+            readOnly: true
+       {{- end }}
+    {{- if .Values.OpenServiceMesh.enableFluentbit}}
```
Comment (shashankram) [inline charts/osm/templates/osm-deployment.yaml:91]:
```suggestion
    {{- if .Values.OpenServiceMesh.enableFluentbit }}
```

File: charts/osm/templates/osm-deployment.yaml [hunk 3]
```diff
@@ -73,6 +73,33 @@ spec:
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.name
+      {{- if .Values.OpenServiceMesh.enableFluentbit}}
+        - name: fluentbit-logger
+          image: fluent/fluent-bit:1.5
+          imagePullPolicy: Always
```
Comment (shashankram) [inline charts/osm/templates/osm-deployment.yaml:79]: IfNotPresent instead of Always?

**Review by shashankram — COMMENTED** [review]
This looks great, some minor questions inline.

File: charts/osm/templates/fluentbit-configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  fluent-bit.conf: |-
+    [SERVICE]
+      Flush             5
+      Daemon            off
+      Log_Level         info
+      Parsers_File      parser.conf
+    [INPUT]
+      Name    tail
+      Tag     kube.*
+      Path    /var/log/containers/osm-controller-*_osm-system_osm-controller-*.log
```
Comment (snehachhabria) [inline charts/osm/templates/fluentbit-configmap.yaml:17]: would this path change if the namespace in which osm is deployed changes ?

File: charts/osm/templates/fluentbit-configmap.yaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+{{- if .Values.OpenServiceMesh.enableFluentbit}}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: fluentbit-configmap
+  namespace: osm-system
```
Comment (ritazh) [inline charts/osm/templates/fluentbit-configmap.yaml:6]: should this be `{{ .Release.Namespace }}`?

File: charts/osm/templates/osm-deployment.yaml [hunk 6]
```diff
@@ -73,6 +73,33 @@ spec:
...
+          volumeMounts:
+          - name: config
+            mountPath: /fluent-bit/etc
+            readOnly: true 
+          - name: var-log-containers
+            mountPath: /var/log/containers
+            readOnly: true
+          - name: var-log-pods
+            mountPath: /var/log/pods
```
Comment (ritazh) [inline charts/osm/templates/osm-deployment.yaml:88]: What component is using this volume?

File: charts/osm/templates/osm-deployment.yaml [hunk 7]
```diff
@@ -73,6 +73,33 @@ spec:
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.name
+      {{- if .Values.OpenServiceMesh.enableFluentbit}}
+        - name: fluentbit-logger
+          image: fluent/fluent-bit:1.5
```
Comment (ritazh) [inline charts/osm/templates/osm-deployment.yaml:78]: Should we also add this as a parameter in values.yaml?

File: cmd/cli/install_test.go [hunk 8]
```diff
@@ -1069,6 +1084,174 @@ var _ = Describe("Ensure that prometheus is disabled when flag is set to false",
...
+		}
+
+		vals, err = installCmd.resolveValues()
+	})
+
+	It("should not error", func() {
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should resolve correctly", func() {
```
Comment (ritazh) [inline cmd/cli/install_test.go:1216]: Can we add a test that inspects the deployed osm-controller pod to make sure the fluentbit  sidecar container exists?

File: docs/development_guide.md [hunk 9]
```diff
@@ -278,6 +278,7 @@ The different chart templates are used as follows:
 - `mutatingwebhook.yaml` is used to deploy a `MutatingWebhookConfiguration` kubernetes resource that enables automatic sidecar injection
 -  `grafana-*.yaml` chart templates are used to deploy a Grafana instance when grafana installation is enabled
 - `prometheus-*.yaml` chart templates are used to deploy a Prometheus instance when prometheus installation is enabled
+- `fluentbit-configmap.yaml` is used to deploy a fluent bit sidecar container with the OSM controller when fluent bit is enabled
```
Comment (ritazh) [inline docs/development_guide.md:281]: Isn't the fluent bit configmap used to provide various configurations to fluent bit?

File: docs/patterns/observability.md [hunk 10]
```diff
@@ -204,3 +204,33 @@ OSM Service to Service Metrics dashboard will look like:
...
 [6]: http://localhost:3000/d/OSMs2sMetrics/osm-service-to-service-metrics?orgId=1
+
+
+# Fluent Bit
+[Fluent Bit](https://fluentbit.io/) is an open source log processor and forwarder which allows you to collect data/logs and send them to multiple destinations. It can be used with OSM to forward OSM controller logs to a variety of outputs/log consumers by using its output plugins.
+
+OSM provides log forwarding by optionally deploying a Fluent Bit sidecar to the OSM controller using the `--enable-fluentbit` flag during installation. The user can then define where OSM logs should be forwarded using any of the available [Fluent Bit output plugins](https://docs.fluentbit.io/manual/v/1.4/pipeline/outputs).
+
+## Configuring Log Forwarding with Fluent Bit
+1. By default, the Fluent Bit sidecar is configured to simply forward logs to an the Fluent Bit container's stdout. Define the output plugin you would like to forward your logs to in the existing `fluentbit-configmap.yaml` file. For example, to forward logs to Azure Monitor, you can use the Azure Log Analytics output plugin by changing the `[OUTPUT]` section as follows:
```
Comment (ritazh) [inline docs/patterns/observability.md:215]: nit: s/to an the/to the

File: charts/osm/templates/fluentbit-configmap.yaml [hunk 11]
```diff
@@ -0,0 +1,34 @@
+{{- if .Values.OpenServiceMesh.enableFluentbit}}
```
Comment (ritazh) [inline charts/osm/templates/fluentbit-configmap.yaml:1]:
```suggestion
{{- if .Values.OpenServiceMesh.enableFluentbit }}
```

File: charts/osm/templates/osm-deployment.yaml (hunk 3 above)
Comment (ritazh) [reply charts/osm/templates/osm-deployment.yaml:79]: A parameter in values.yaml?

File: charts/osm/templates/osm-deployment.yaml (hunk 7 above)
Comment (SanyaKochhar) [reply charts/osm/templates/osm-deployment.yaml:78]: by this do you mean the name of the sidecar?

Comment (ritazh) [reply charts/osm/templates/osm-deployment.yaml:78]: I'm referring to the image name and tag. Similar to how users can configure the values for the osm controller image by changing the values in the values.yaml:
`image: "{{ .Values.OpenServiceMesh.image.registry }}/osm-controller:{{ .Values.OpenServiceMesh.image.tag }}"`

Comment (SanyaKochhar) [reply charts/osm/templates/osm-deployment.yaml:78]: ah yes, that makes sense

File: charts/osm/templates/fluentbit-configmap.yaml (hunk 4 above)
Comment (SanyaKochhar) [reply charts/osm/templates/fluentbit-configmap.yaml:17]: yes, good catch!

File: charts/osm/templates/osm-deployment.yaml (hunk 6 above)
Comment (SanyaKochhar) [reply charts/osm/templates/osm-deployment.yaml:88]: Fluent Bit uses these to get access to the osm-controller logs that by default are stored in var/log/containers/. I'm not totally sure why var/log/pods is needed but it does not find any logs without it

File: cmd/cli/install_test.go (hunk 8 above)
Comment (SanyaKochhar) [reply cmd/cli/install_test.go:1217]: Yes I am working on this

File: tests/e2e/e2e_fluentbit_test.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
+package e2e
```
Comment (eduser25) [inline tests/e2e/e2e_fluentbit_test.go:1]: @shashankram  @nojnhuh , we need to cut down the CI and gating tests, we should meet to draw the line. 
Additionally we have some additional concerns regarding parallelization, gh actions might run oversubscribed with multiple kind clusters.

I'll schedule a meeting to agree on path forward, threshold of acceptable CI window time, etc.

**Review by eduser25 — APPROVED** [review]
lgtm

File: charts/osm/templates/osm-deployment.yaml (hunk 6 above)
Comment (ritazh) [reply charts/osm/templates/osm-deployment.yaml:88]: Ah that makes sense. I did a quick search and found `Kubernetes is using symbolic links from /var/log/containers to /var/log/pods` ref: https://github.com/kubernetes/kubernetes/issues/53022

**Review by ritazh — APPROVED** [review]
lgtm

File: charts/osm/templates/osm-deployment.yaml (hunk 6 above)
Comment (SanyaKochhar) [reply charts/osm/templates/osm-deployment.yaml:88]: oh that explains why some fluentbit examples had var/lib/containers instead! Can I mark this as resolved?

File: docs/patterns/observability.md [hunk 13]
```diff
@@ -204,3 +204,33 @@ OSM Service to Service Metrics dashboard will look like:
...
+OSM provides log forwarding by optionally deploying a Fluent Bit sidecar to the OSM controller using the `--enable-fluentbit` flag during installation. The user can then define where OSM logs should be forwarded using any of the available [Fluent Bit output plugins](https://docs.fluentbit.io/manual/v/1.4/pipeline/outputs).
+
+## Configuring Log Forwarding with Fluent Bit
+1. By default, the Fluent Bit sidecar is configured to simply forward logs to the Fluent Bit container's stdout. Define the output plugin you would like to forward your logs to in the existing `fluentbit-configmap.yaml` file. For example, to forward logs to Azure Monitor, you can use the Azure Log Analytics output plugin by changing the `[OUTPUT]` section as follows:
+   ```
+   [OUTPUT]
+         Name        azure
+         Match       *
+         Customer_ID <Log Analytics Workspace ID>
+         Shared_Key  <Log Analytics Primary key>
```
Comment (michelleN) [inline docs/patterns/observability.md:221]: Is this a secret?

**Pull request merged**